// view/BaseInstrument.js
var inherit = require("../utils/inherit");
var BaseView = require("./BaseView");

module.exports = BaseInstrument;

function BaseInstrument(track, instrumentName){
    BaseView.call(this, instrumentName);
    this.track = track;
    this.startTime = 0;
    this.isRecordNow = false;
    this.isActive = false;
}

inherit(BaseInstrument, BaseView);

BaseInstrument.prototype.recordEvent = function(eventName, recordButton){
    if(this.isRecordNow === true){
        // stop record
        this.isRecordNow = false;
        recordButton.text("record");
        recordButton.attr("data-content", "Click to record sound");
    } else{
        this.isRecordNow = true;
        this.track.emptyPlaySetting(); // clear previous play data setting
        recordButton.text("stop");
        recordButton.attr("data-content", "Click to stop record");
    }
};

BaseInstrument.prototype.clearEvent = function(){
    this.track.emptyPlaySetting();
};

BaseInstrument.prototype.keyDown = function(){};

BaseInstrument.prototype.keyUp = function(){};
