// view/ContentView.js
var inherit = require("../utils/inherit");
var BaseView = require("./BaseView");
var ProjectList = require("./ProjectList");
var eventListener = require("../model/eventListener");

module.exports = ContentView;

function ContentView(){
    BaseView.call(this, "content-view");

    this.projectList = new ProjectList(["teeeeest1", "test2", "sibsneaepinb", "qqqqqq",
        "rr", "rwythrjtejedbxd", "rrr", "wegrfgnh",
        "aknv", "teeeeest1", "test2", "sibsneaepinb", "qqqqqq",
        "rr", "rwythrjtejedbxd", "rrr", "wegrfgnh",
        "aknv"]);

    this.sideBar = null;
    this.trackView = null;

    this._build();
}

inherit(ContentView, BaseView);

ContentView.prototype._build = function(){
    var container = this.getContainer();

    if (!eventListener.subscribe(eventListener.ON_SHOW_PROJECT_LIST, this.showProjectList.bind(this))){

        console.error("Unable to subscribe to", eventListener.ON_SHOW_PROJECT_LIST, "event!");
    }
    // EXAMPLE:
    eventListener.subscribe("ADD_TRACK", this.addTrack.bind(this));

    container.append(this.projectList);

    this.appendToBlock($(".ui.container"));
};

ContentView.prototype.showProjectList = function(){
    this.hideALl();
    this.projectList.show();
};

ContentView.prototype.hideALl = function(){
    this.projectList.hide();
    // ... hide all evelements
};

ContentView.prototype.addTrack = function(trackInfo){
    this.sideBar.addTrack(trackInfo);
    this.trackView.addTrack(trackInfo);
};
