// view/ContentView.js
var inherit = require("../utils/inherit");
var BaseView = require("./BaseView");
var ProjectList = require("./ProjectList");
var TrackToolView = require("./TrackToolView");
var TrackView = require("./WaveForm");
var eventListener = require("../model/eventListener");

module.exports = ContentView;

function ContentView(){
    BaseView.call(this, "content-view");

    this.projectList = new ProjectList();
    this.trackToolView = new TrackToolView();
    this.trackView = new TrackView();

    this.sideBar = null;
    //this.trackView = null;

    this._build();
}

inherit(ContentView, BaseView);

ContentView.prototype._build = function(){
    var container = this.getContainer();

    /*if (!eventListener.subscribe(eventListener.ON_SHOW_PROJECT_LIST, this.showProjectList.bind(this))){

        console.error("Unable to subscribe to", eventListener.ON_SHOW_PROJECT_LIST, "event!");
    }*/
    // EXAMPLE:

    eventListener.notify("SHOW_TRACK");

    eventListener.subscribe("ON_SHOW_PROJECT_LIST", this.showProjectList.bind(this));
    eventListener.subscribe("ON_ADD_PROJECT", this.showTrackToolView.bind(this));
    eventListener.subscribe("ADD_TRACK", this.addTrack.bind(this));

    container.append(this.projectList.getContainer());
    container.append(this.trackToolView.getContainer());
    container.append(this.trackView.getContainer());

    this.appendToBlock($(".ui.container"));
};

ContentView.prototype.showProjectList = function(){
    this.hideAll();
    this.getContainer().append(this.projectList.getContainer());
    this.projectList.fetchProjectList();
    this.projectList.show();
};

ContentView.prototype.showTrackToolView = function(){
    this.hideAll();
    this.trackToolView.show();
};

ContentView.prototype.hideAll = function(){
    this.projectList.getContainer().remove();
    //this.projectList.hide();
    this.trackToolView.hide();
    // ... hide all evelements
};

ContentView.prototype.addTrack = function(trackInfo){
    this.sideBar.addTrack(trackInfo);
    this.trackView.addTrack(trackInfo);
};
