// view/Factory.js
module.exports = {
    "buttonsPopup": buttonsPopup,
    "createButton": createButton,
    "createDivButton": createDivButton,
    "createIconButton": createIconButton,
    "deleteCircleButton": deleteCircleButton
};

function createButton(className, text){
    return $("<button class='ui button " + className + "'>" + text + "</button>");
}

function createDivButton(className, text){
    return $("<div class='" + className + "'>" + text + "</div>");
}

function createIconButton(buttonClass, iconClass, name){
    return $("<button class='" + buttonClass + "'>" +
        "<i class='" + iconClass + "'></i>" + name + "</button>");
}

function deleteCircleButton(buttonId, callback){
    var $deleteProjectButton = createIconButton("circular ui icon button", "remove icon", "");
    $deleteProjectButton.attr("id", buttonId);
    $deleteProjectButton.on("click", function(event){
        callback($(this));
    });
    return $deleteProjectButton;
}

/**
 * Create button set with the same callback
 * @param buttonNameList
 * @param callback
 */
function buttonsPopup(buttonNameList, callback){
    var i;
    var $column;
    var $button;
    var $result = $("<div class='ui flowing popup top left transition hidden'>");
    var $row = $("<div class='ui three column divided center aligned grid'>");
    for(i = 0; i < buttonNameList.length; ++i){
        $column = $("<div class='column'>");
        $button = createButton("", buttonNameList[i]);
        $button.on("click", function(event){
            callback($(this).val());
        });
        $column.append($button);
        $row.append($column);
    }
    $result.append($row);
    return $result;
}

function createPiano(){
    var $result = $("<div class='piano'>");

}
