// view/FilterView.js
var inherit = require("../utils/inherit");
var ToolView = require("./ToolView");
var FiltersList = require("../model/FiltersList");
var ProxyTrackManager = require("../model/ProxyTrackManager");
var Factory = require("./Factory");

module.exports = FilterView;

function FilterView(filter){
    ToolView.call(this, "filter-view", "two column stackable ui grid");

    this.filter = null;

    this.setFilter(filter);

    this._build();
    this.show();
}

inherit(FilterView, ToolView);


FilterView.prototype.setFilter = function(track){
    if(track){
        this.filter = track.filterObject;
        ProxyTrackManager.updateFilterListFromFilter(FiltersList, this.filter.postSettings);
        this.createFilterTools();
    }
};

FilterView.prototype.createFilterTools = function(){
    var i;
    var value;
    var options;
    var filterName;
    var optionName;
    var tokenFilter;
    var $element;
    var $elementName;
    var $mainDiv;
    var $checkBox;
    var $settingDiv;
    for(i = 0; i < FiltersList.list.length; ++i){
        tokenFilter = FiltersList.list[i];
        options = tokenFilter.options;
        filterName = tokenFilter.name;
        $mainDiv = $("<div class='column " + filterName + "'>");
        for(optionName in options){
            value = options[optionName]; // BaseOption
            $settingDiv = $("<div class='filter-setting'>");
            $elementName = $("<label>" + optionName + "</label>");
            $element = this.createElement(optionName, value);
            $settingDiv.append($elementName);
            $settingDiv.append($element);
            $settingDiv.hide();
        }
        $checkBox = Factory.createCheckBox(filterName, filterName, filterName, function(name){
            checkEvent($settingDiv);
        }, function(name){
            uncheckEvent($settingDiv);
        }, tokenFilter.isEnabled);
        $mainDiv.append($checkBox);
        $mainDiv.append($settingDiv);
        this.table.append($mainDiv);
    }
};

FilterView.prototype.setEvent = function(optionName, value){
    console.log(this.track);
    ProxyTrackManager.setSetting(FiltersList, this.filter, optionName, value);
};

function checkEvent(settingDiv){
    console.log("+");
    settingDiv.show();
}

function uncheckEvent(settingDiv) {
    settingDiv.hide();
}
