// view/FilterView.js
var inherit = require("../utils/inherit");
var ToolView = require("./ToolView");
var FiltersList = require("../model/FiltersList");
var ProxyTrackManager = require("../model/ProxyTrackManager");
var Factory = require("./Factory");

module.exports = FilterView;

function FilterView(filter){
    ToolView.call(this, "filter-view", "two column stackable ui grid");

    this.filter = null;

    this.setFilter(filter);

    this._build();
    this.show();
}

inherit(FilterView, ToolView);

FilterView.prototype.resetFilters = function(){
    ProxyTrackManager.resetFilters(FiltersList);
};

FilterView.prototype.setFilter = function(track){
    if(track){
        this.filter = track.postSettings;
        ProxyTrackManager.updateFilterListFromFilter(FiltersList, this.filter.getPostSettings());
        this.createFilterColumns();
    }
};

FilterView.prototype.setEvent = function(optionName, value){
    ProxyTrackManager.setFilter(FiltersList, this.filter, optionName, value);
};

/**
 * Full div with filter setting of rangeElement || dropDownElement
 */
FilterView.prototype.createFilterSettingElements = function(mainBlock, options){
    var value;
    var optionName;
    var $element;
    var $elementName;
    for(optionName in options){
        value = options[optionName]; // BaseOption
        $elementName = $("<label>" + optionName + "</label>");
        $element = this.createElement(optionName, optionName, value);
        mainBlock.append($elementName);
        mainBlock.append($element);
    }
};

FilterView.prototype.createFilterTools = function(count, startValue){
    var i = startValue || 0;
    var options;
    var filterName;
    var tokenFilter;
    var $mainDiv;
    var $checkBox;
    var $settingDiv;
    var $result = $("<div>");
    for(i; i < count; ++i){
        tokenFilter = FiltersList.list[i];
        options = tokenFilter.options;
        filterName = tokenFilter.name;
        // create main div for checkbox component and div with settings
        $mainDiv = $("<div class='item-" + filterName + "'>");
        $settingDiv = $("<div class='filter-setting-" + filterName + "'>");
        // create widgets for settings
        this.createFilterSettingElements($settingDiv, options);
        if(tokenFilter.isEnabled === true){
            $settingDiv.show();
        } else {
            $settingDiv.hide();
        }
        $checkBox = Factory.createCheckBox(filterName, filterName, filterName, checkEvent.bind(this),
                                           uncheckEvent.bind(this), tokenFilter.isEnabled);
        $mainDiv.append($checkBox);
        $mainDiv.append($settingDiv);
        $result.append($mainDiv);
    }
    return $result;
};

FilterView.prototype.createFilterColumns = function(){
    var $column1;
    var $column2;
    var elCountFirstCol = Math.ceil(FiltersList.list.length / 2);
    $column1 = this.createFilterTools(elCountFirstCol);
    $column2 = this.createFilterTools(FiltersList.list.length, elCountFirstCol);
    this.table.append($column1);
    this.table.append($column2);
};

function checkEvent(filterName, isChecked){
    this.table.find(".filter-setting-" + filterName).show();
    ProxyTrackManager.addFilter(FiltersList, this.filter, filterName, isChecked);
}

function uncheckEvent(filterName, isChecked){
    this.table.find(".filter-setting-" + filterName).hide();
    ProxyTrackManager.addFilter(FiltersList, this.filter, filterName, isChecked);
}
