// view/MenuBar.js
var inherit = require("../utils/inherit");
var BaseView = require("./BaseView");
var UserInfoBar = require("./UserInfoBar");
var TrackView = require("./TrackView");
var TrackListView = require("./TrackListView");
var ProjectListView = require("./ProjectListView");
var PlayerView = require("./PlayerView");
var Factory = require("./Factory");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("./windowsTransport");


module.exports = MenuBar;

function MenuBar(){
    BaseView.call(this, "menubar");

    this.userInfoBar = new UserInfoBar();
    this.backButton = Factory.createIconButton("ui button back user-only", "arrow left icon", "");
    this.player = new PlayerView();
    this.exportProjectButton = Factory.createButton("export-project", "export WAV");
    this.exportTrackButton = Factory.createButton("export-track", "export WAV");
    this.recordButton = Factory.createButton("record", "record");
    this.clearButton = Factory.createButton("clear", "clear");

    this._build();
}

inherit(MenuBar, BaseView);

MenuBar.prototype._build = function(){
    var container = this.getContainer();

    //windowsTransport.notify(commonEventNames.E_DEFINE_USER);

    this.backButton.on("click", function(event){
        // Possible additional visual effect for this button...
        windowsTransport.notify(commonEventNames.ON_BACK_BUTTON_CLICK);
    });

    this.recordButton.on("click", function(event){
        windowsTransport.notify(commonEventNames.E_RECORD_TRACK, $(this));
    });

    this.clearButton.on("click", function(event){
        windowsTransport.notify(commonEventNames.E_CLEAR_RECORD_TRACK);
    });

    this.hideButtons();

    container.append(this.backButton);
    container.append(this.player.getContainer());
    container.append(this.exportProjectButton);
    container.append(this.exportTrackButton);
    container.append(this.recordButton);
    container.append(this.clearButton);
    container.append(this.userInfoBar.getContainer());
};

MenuBar.prototype.adaptToActiveWindow = function(newWindow){
    if(newWindow instanceof TrackView){
        this.showComponentForTrack(newWindow);
        this.recordButton.show();
        this.clearButton.show();
        this.exportTrackButton.show();
        this.exportProjectButton.hide();

        this.exportTrackButton.on("click", function(event){
            console.log("export");
            console.log(newWindow);
            windowsTransport.notify(commonEventNames.E_EXPORT_TRACK);
        });
    } else if(newWindow instanceof TrackListView){
        this.showComponentForTrack(newWindow);
        this.hideButtons();
        this.exportProjectButton.show();

        this.exportProjectButton.on("click", function(event){
            console.log("export");
            console.log(newWindow);
            windowsTransport.notify(commonEventNames.E_EXPORT_PROJECT);
        });
    } else if(newWindow instanceof ProjectListView){
        this.backButton.hide();
        this.player.hide();
        this.exportProjectButton.hide();
    }
};

MenuBar.prototype.showComponentForTrack = function(newWindow){
    this.backButton.show();
    this.player.show();
    this.player.audioPlayer.setModel(newWindow.controller.model);
};

MenuBar.prototype.hideButtons = function(){
    this.exportTrackButton.hide();
    this.recordButton.hide();
    this.clearButton.hide();
};
