// view/MenuBar.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var ProjectBar = require("./ProjectBar");
var UserModal = require("./UserModal");

module.exports = MenuBar;

function MenuBar(){
    BaseView.call(this, "menubar");

    this.projecrBar = null;
    this.userInfo = $("<div class='user-info'></div>");

    this.userName = null;
    this.joinButton = Factory.createButton("join");

    this.joinButton.on("click", function(){
        var userModal = new UserModal();
        userModal.show();
    });

    this._build();
    this.appendToBlock(".ui.container");
}

inherit(MenuBar, BaseView);

MenuBar.prototype._build = function(){
    this.userInfo.append(this.userName);  // empty
    this.userInfo.append(this.joinButton);
    this._container.append(this.projecrBar); // empty
    this._container.append(this.userInfo);
};

MenuBar.prototype.showMenuComponents = function(userName){
    if(userName instanceof Error){
        this.userName = $("<p>anonym</p>");
    } else{
        this.projecrBar = new ProjectBar(this._container);
        this.userName = $("<p>" + userName + "</p>");
        this._container.append(this.projecrBar);
    }
    this.userInfo.append(this.userName);
};
