// view/Oscillator.js
var inherit = require("../utils/inherit");
var DraggerPlayer = require("../model/DraggerPlayer");
var BaseView = require("./BaseView");

module.exports = Oscillator;

function Oscillator(track){
    BaseView.call(this, "oscillator");
    this.track = track;
    this.draggable = null;
    this.draggie = null;

    this.startTime = 0;
    this.isRecordNow = false;
    this.isMouseClicked = false;

    this.coordsMap = $("<div class='coord-map'>");
    this.oscillator = $("<div class='circle-drag draggable'>");
    this._build();
}

inherit(Oscillator, BaseView);

Oscillator.prototype._build = function(){
    var self = this;
    var container = this.getContainer();

    this.draggable = this.oscillator.draggabilly();
    this.draggie = this.draggable.data("draggabilly");

    setPosition(this.draggie, this.track.getFrequency(), this.track.getVolume());

    this.draggable.on("dragMove", function(event){
        self.dragMoveEvent();
    });

    this.draggable.on("pointerDown", function(event){
        self.dragMoveEvent();
    });

    this.draggable.on("dragEnd", function(event){
        self.dragEndEvent();
    });

    this.coordsMap.append(this.oscillator);
    container.append(this.coordsMap);
};

function setPosition(draggieObject, x, y) {
    draggieObject.position.x = x;
    draggieObject.position.y = y;
    draggieObject.setLeftTop();
}

Oscillator.prototype.dragMoveEvent = function(){
    this.track.setFrequency(this.draggie.position.x);
    this.track.setVolume(this.draggie.position.y);
    this.track.trackObject.start();
    console.log( 'dragMove', this.draggie.position.x, this.draggie.position.y );
    if(this.track.playObjects.length === 0){
        this.startTime = Date.now();
        this.track.playObjects.push(new DraggerPlayer(this.draggie.position.x, this.draggie.position.y, 0));
    } else{
        this.track.playObjects.push(new DraggerPlayer(this.draggie.position.x, this.draggie.position.y,
                                                      Date.now()-this.startTime));
    }
};

Oscillator.prototype.dragEndEvent = function(){
    this.track.trackObject.stop();
};
