// view/PlayButtonBar.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var eventListener = require("../model/eventListener");

module.exports = PlayButtonBar;

function PlayButtonBar(){
    BaseView.call(this, "play-button-group");

    this.playButton = Factory.createIconButton("ui button", "play icon", "");
    this.pauseButton = Factory.createIconButton("ui button", "pause icon", "");
    this.stopButton = Factory.createIconButton("ui button", "stop icon", "");
    this.prevButton = Factory.createIconButton("ui button", "step backward icon", "");
    this.nextButton = Factory.createIconButton("ui button", "step forward icon", "");

    this._build();
}

inherit(PlayButtonBar, BaseView);

PlayButtonBar.prototype._build = function(){
    var container = this.getContainer();

    this.playButton.on("click", function(_){
    });
    this.pauseButton.on("click", function(_){
    });
    this.stopButton.on("click", function(_){
    });
    this.prevButton.on("click", function(_){
    });
    this.nextButton.on("click", function(_){
    });

    container.append(this.prevButton);
    container.append(this.playButton);
    container.append(this.pauseButton);
    container.append(this.stopButton);
    container.append(this.nextButton);
};
