// view/ProjectBar.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var ProjectList = require("./ProjectList");
var TrackToolView = require("./TrackToolView");

module.exports = ProjectBar;

function ProjectBar(bottomContainer, contentContainer, projectList){
    BaseView.call(this, "projectbar");

    this.contentContainer = contentContainer;

    this.editingProjectButton = $("<div class='editing-project-button'></div>");

    this.showProjectListButton = Factory.createIconButton("ui button", "book icon", "");
    this.addButton = Factory.createIconButton("ui button", "plus icon", "");
    this.editButton = Factory.createIconButton("ui button", "disabled edit icon", "");
    this.deleteButton = Factory.createIconButton("ui button", "disabled trash icon", "");

    this.isPressedProjectList = true;
    this.isPressedAddProject = true;

    this.projectList = new ProjectList(["teeeeest1", "test2", "sibsneaepinb", "qqqqqq",
        "rr", "rwythrjtejedbxd", "rrr", "wegrfgnh",
        "aknv", "teeeeest1", "test2", "sibsneaepinb", "qqqqqq",
        "rr", "rwythrjtejedbxd", "rrr", "wegrfgnh",
        "aknv"]);

    this.trackToolView = new TrackToolView();

    this._build(bottomContainer);

    this.projectList.hide();
    this.trackToolView.hide();
}

inherit(ProjectBar, BaseView);

ProjectBar.prototype._build = function(bottomContainer){
    var self = this;

    this.showProjectListButton.on("click", function(event){
        if(self.isPressedProjectList) {
            self.projectList.show();
            self.contentContainer.hide();
            self.isPressedProjectList = false;
        } else{
            self.projectList.hide();
            self.contentContainer.show();
            self.isPressedProjectList = true;
        }
    });

    this.addButton.on("click", function(event){
        if(self.isPressedAddProject){
            self.trackToolView.show();
            self.contentContainer.hide();
            self.isPressedAddProject = false;
        } else{
            self.trackToolView.hide();
            self.contentContainer.show();
            self.isPressedAddProject = true;
        }
        console.log("add");
    });

    this.editButton.on("click", function(event){

    });

    this.deleteButton.on("click", function(event){
        var projectName = "project_DELETE";  // test
        //RequestManager.deleteProject(projectName);
    });

    this._container.append(this.addButton);
    this.editingProjectButton.append(this.editButton);
    this.editingProjectButton.append(this.deleteButton);
    this.editingProjectButton.append(this.showProjectListButton);
    this._container.append(this.editingProjectButton);
    bottomContainer.append(this._container);
};

ProjectBar.prototype.hideEditingProjectButton = function(){
    this.editingProjectButton.hide();
};
