// view/ProjectBaseView.js
var inherit = require("../utils/inherit");
var BaseView = require("./BaseView");
var MenuBar = require("./MenuBar");
var ContentView = require("./ContentView");
var PlayerView = require("./PlayerView");
var RequestManager = require("../RequestManager");

module.exports = ProjectBaseView;

function ProjectBaseView(){
    BaseView.call(this, "base-view");

    this.menuBar = new MenuBar();
    this.contentView = new ContentView();
    this.player = new PlayerView();

    this._build();
    this.appendToBlock($(".ui.container"));
}

inherit(ProjectBaseView, BaseView);

ProjectBaseView.prototype._build = function(){
    var container = this.getContainer();

    var requestManager = new RequestManager();
    requestManager.getUser(this.hideUserOnlyElements.bind(this));

    container.append(this.menuBar.getContainer());
    container.append(this.contentView.getContainer());
    container.append(this.player.getContainer());

    // eventListener.subscribe("ON_SHOW_PROJECT_LIST", this._fetchProjectList.bind(this));
};

ProjectBaseView.prototype.hideUserOnlyElements = function(userName){
    if(userName instanceof Error){
        var $elements = $(".user-only");
        $elements.addClass("disabled");
        // ... add "disable" class to all of these $elements
    }
};
