// view/ProjectView.js
// ProjectView
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var ProjectListFactory = require("./Factory");
var RequestManager = require("../RequestManager");


module.exports = ProjectView;

function ProjectView(){
    BaseView.call(this, "project-bar");
    this.projectList = $("<div class='project-list'></div>");
    this.projecManager = $("<div class='project-manager'></div>");

    this.list = null;

    this.addButton = Factory.createIconButton("ui button", "plus icon", "");
    this.editButton = Factory.createIconButton("ui button", "disabled edit icon", "");
    this.deleteButton = Factory.createIconButton("ui button", "disabled trash icon", "");

    this.addButton.on("click", function(event){
        console.log("add");
    });

    this.editButton.on("click", function(event){

    });

    this.deleteButton.on("click", function(event){
        var projectName = "project_DELETE";  // test
        //RequestManager.deleteProject(projectName);
    });
}

inherit(ProjectView, BaseView);

ProjectView.prototype.fullProjectList = function(data){
    this.list = ProjectListFactory.createList(data);
    this._build();
};

ProjectView.prototype._build = function(){
    this.projectList.append(this.list);
    this.projecManager.append(this.addButton);
    this.projecManager.append(this.editButton);
    this.projecManager.append(this.deleteButton);
    this._container.append(this.projectList);
    this._container.append(this.projecManager);
};
