// view/ProjectBar.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var ProjectList = require("./ProjectList");

module.exports = ProjectBar;

function ProjectBar(bottomContainer){
    BaseView.call(this, "projectbar");

    this.showProjectListButton = Factory.createIconButton("ui button", "book icon", "");
    this.addButton = Factory.createIconButton("ui button", "plus icon", "");
    this.editButton = Factory.createIconButton("ui button", "disabled edit icon", "");
    this.deleteButton = Factory.createIconButton("ui button", "disabled trash icon", "");

    this.isPressedProjectList = true;
    this.projectList = new ProjectList(["teeeeest1", "test2", "sibsneaepinb", "qqqqqq",
        "rr", "rwythrjtejedbxd", "rrr", "wegrfgnh",
        "aknv", "teeeeest1", "test2", "sibsneaepinb", "qqqqqq",
        "rr", "rwythrjtejedbxd", "rrr", "wegrfgnh",
        "aknv"]);

    this._build(bottomContainer);

    this.projectList.hide();
}

inherit(ProjectBar, BaseView);

ProjectBar.prototype._build = function(bottomContainer){
    var self = this;

    this.showProjectListButton.on("click", function(event){
        if(self.isPressedProjectList) {
            self.projectList.show();
            self.isPressedProjectList = false;
        } else{
            self.projectList.hide();
            self.isPressedProjectList = true;
        }
    });

    this.addButton.on("click", function(event){
        console.log("add");
        this.projectList.show();
    });

    this.editButton.on("click", function(event){

    });

    this.deleteButton.on("click", function(event){
        var projectName = "project_DELETE";  // test
        //RequestManager.deleteProject(projectName);
    });

    this._container.append(this.showProjectListButton);
    this._container.append(this.addButton);
    this._container.append(this.editButton);
    this._container.append(this.deleteButton);
    bottomContainer.append(this._container);
};
