// view/ProjectView.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var RequestManager = require("../RequestManager");


module.exports = ProjectView;

function ProjectView(){
    BaseView.call(this, "project-bar");
    this.projectListBlock = $("<div class='project-list'></div>");
    this.projecManager = $("<div class='project-manager'></div>");

    this.projectList = null;

    this.addButton = Factory.createIconButton("ui button", "plus icon", "");
    this.editButton = Factory.createIconButton("ui button", "disabled edit icon", "");
    this.deleteButton = Factory.createIconButton("ui button", "disabled trash icon", "");

    this.addButton.on("click", function(event){
        console.log("add");
    });

    this.editButton.on("click", function(event){

    });

    this.deleteButton.on("click", function(event){
        var projectName = "project_DELETE";  // test
        //RequestManager.deleteProject(projectName);
    });

    this._build();
    this.appendToBlock(".ui.container");
}

inherit(ProjectView, BaseView);

ProjectView.prototype._build = function(){
    this.projectListBlock.append(this.projectList);
    this.projecManager.append(this.addButton);
    this.projecManager.append(this.editButton);
    this.projecManager.append(this.deleteButton);
    this._container.append(this.projectListBlock);
    this._container.append(this.projecManager);
};

ProjectView.prototype.fullProjectList = function(data){
    console.log(this);
    this.projectList = Factory.createList(data);

    this.projectListBlock.append(this.projectList);
};
