// view/ProjectList.js
var inherit = require("../utils/inherit");
var BaseWindow = require("./BaseWindow");
var Factory = require("./Factory");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("./windowsTransport");

module.exports = ProjectList;

function ProjectList(controller){
    BaseWindow.call(this, controller, "project-list");

    this.title = $("<h1>Project List</h1>");
    this.table = $("<div class='five column stackable ui grid'>");
    this.addProjectButton = Factory.createIconButton("circular ui icon button", "plus icon", "");

    this.selectedItem = null;
    this.onRemoveButtonClicked = onRemoveButtonClicked.bind(this);

    this._build();
    this.hide();
}

inherit(ProjectList, BaseWindow);

ProjectList.prototype._build = function(){
    var self = this;
    var container = this.getContainer();

    this.controller.observer.subscribe(commonEventNames.E_ITEM_ADDED, function(eventName, index){
        var trackListModel = self.controller.model.at(index);
        self.add(trackListModel);
    });

    this.controller.observer.subscribe(commonEventNames.E_ITEM_REMOVED, function(eventName, index){
        self.remove(self.selectedItem.attr("id"));
        self.selectedItem = null;
    });

    this.addProjectButton.on("click", function(event){
        self.controller.add(JSON.stringify({"DEFAULT_KEY": "DEFAULT_VALUE"}));
    });

    container.append(this.title);
    container.append(this.table);
    container.append(this.addProjectButton);
};

ProjectList.prototype.confirmed = function(name){
    this.controller.remove(this.selectedItem.attr("id"));
    //this.selectedItem = null;
};

ProjectList.prototype.declined = function(name){
    this.selectedItem = null;
};

ProjectList.prototype.add = function(name){
    // full table or
    // add new item with new project
    var $item;
    var $column;
    var projectName;
    var $deleteProjectButton;
    $deleteProjectButton = Factory.deleteCircleButton(name, this.onRemoveButtonClicked);
    //create function for project name ITEM
    $item = $("<a id='" + name + "'>" + name +"</a>");
    $item.on("click", function(event){
        projectName = $(this).attr("id");
        windowsTransport.notify(commonEventNames.SHOW_PROJECT, projectName);
    });
    $column = $("<div class='column' id='item-" + name + "'>");
    $column.append($item);
    $column.append($deleteProjectButton);
    this.table.append($column);
};

ProjectList.prototype.remove = function(name){
    $("#item-" + name).remove(); //remove from table (grid)
};

ProjectList.prototype.updateProjectList = function(){
    this.table.empty();
    this.controller.fetchData();
};

function onRemoveButtonClicked($element){
    if (this.selectedItem === null){
        this.selectedItem = $element;
        windowsTransport.notify(commonEventNames.E_SHOW_MODAL, "teeest");
    }
}
