// view/TrackView.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var RequestManager = require("../RequestManager");
var eventListener = require("../model/eventListener");

module.exports = TrackView;

function TrackView(){
    BaseView.call(this, "track-view");

    this.waveContainer = $("<div id='waveform'></div>");
    this.waveform = null;

    this._build();
}

inherit(TrackView, BaseView);

TrackView.prototype._build = function(){
    eventListener.subscribe("SHOW_TRACK", this._fetchSoundData.bind(this));

    this.getContainer().append(this.waveContainer);
};

TrackView.prototype._fetchSoundData = function(){
    var requestManager = new RequestManager();
    requestManager.uploadSound("test.wav", this.createWaveForm.bind(this));
};

TrackView.prototype.createWaveForm = function(blob){
    this.waveform = WaveSurfer.create({
        container: "#waveform",
        waveColor: "#626262",
        progressColor: "#fff843"
    });
    this.waveform.loadBlob(blob);
};
