// view/TrackView.js
//TrackComponentView
var inherit = require("../utils/inherit");
var AudioHelper = require("../utils/AudioHelper");
var TrackManager = require("../utils/TrackManager");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var WaveForm = require("./WaveForm");
var InstrumentView = require("./InstrumentView");
var eventListener = require("../model/eventListener");

module.exports = TrackView;

function TrackView(trackModel){
    BaseView.call(this, "track-view");

    this.trackModel = trackModel;

    this.instrumentView = null;
    this.waveform = null;
    this.deleteButton = Factory.deleteCircleButton(this.trackModel.name, null);

    this._build(trackModel);
}

inherit(TrackView, BaseView);

TrackView.prototype._build = function(trackModel){
    //eventListener.subscribe("SHOW_TRACK", this._fetchSoundData.bind(this));
    //eventListener.notify("SHOW_TRACK");
    this.getContainer().append(this.deleteButton);
    this.createInstrument();
    this.createWaveForm();
};

TrackView.prototype.createInstrument = function(){
    this.instrumentView = new InstrumentView(this.trackModel.name, this.trackModel.instrument);
    this.getContainer().append(this.instrumentView.getContainer());
};

TrackView.prototype.createWaveForm = function(){
    /*var data = this.trackModel.getContext();
    //this.trackModel.play();
    //this.trackModel.saveTest(TrackManager.save);
    this.waveform = new WaveForm();
    this.getContainer().append(this.waveform.getContainer());
    if(data instanceof Blob){
        this.waveform.createWaveFormFromFile(data);
    } else if(data instanceof AudioContext){
        this.waveform.createWaveForm(data);
    }*/
    var self = this;
    Tone.Offline(function(){
        //only nodes created in this callback will be recorded
        //var oscillator = new Tone.Oscillator().toMaster().start(0);
        var synthBass = new Tone.Synth({"oscillator": {"frequency": 440},
            "envelope": {"attack" : 0.5, "decay": 0.9, "sustain": 0.3, "release": 1}}).toMaster();
        synthBass.triggerAttackRelease('C4', 5);
        synthBass.triggerAttackRelease('A2', 2, 2);
        synthBass.triggerAttackRelease('A2', 2, 3);
        //schedule their events
    }, 5).then(function(buffer){
        //do something with the output buffer
        console.log(buffer);
        //TrackManager.save(buffer._buffer);
        var blob = AudioHelper.AudioBufferToBlob(buffer._buffer);
        self.waveform = new WaveForm();
        self.getContainer().append(self.waveform.getContainer());
        self.waveform.createWaveFormFromFile(blob);

        /*
        self.waveform = new WaveForm();
        self.getContainer().append(self.waveform.getContainer());

        var p = new Tone.Player(buffer._buffer).toMaster().start();
        self.waveform.createWaveForm(p.context._context);
         */
    })

};
