// view/UserModal.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");

module.exports = UserModal;

function UserModal(){
    BaseView.call(this, "ui modal");
    this.isJoin = false;
    this.facebookLogIn = Factory.createIconButton("ui facebook button", "facebook icon", "Facebook");
    this.label = $("<div class='ui pointing below label'>Join without registration");
    this.button = Factory.createButton("join");
    this.buttonLogout = Factory.createButton("Logout");

    this.facebookLogIn.on("click", function(event){
        location.href = "/accounts/facebook/login/";
        console.log("q");
        //RequestManager.projectList(ProjectView.fullProjectList);
    });

    this.buttonLogout.on("click", function(event){
        location.href = "/accounts/logout/";
        console.log("logout");
    });

    this._build();
}

inherit(UserModal, BaseView);

UserModal.prototype._build = function(){
    this._container.append(this.facebookLogIn);
    this._container.append(this.label);
    this._container.append(this.button);
    this._container.append(this.buttonLogout);
};

UserModal.prototype.show = function(){
    this._container.modal("show");
};
