// view/WaveForm.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var RequestManager = require("../RequestManager");
var eventListener = require("../model/eventListener");

module.exports = WaveForm;

function WaveForm(){
    BaseView.call(this, "waveform");

    $(this.getContainer()).attr("id", "waveform");

    this.waveform = null;

    this._build();
}

inherit(WaveForm, BaseView);

WaveForm.prototype._build = function(){
    eventListener.subscribe("SHOW_TRACK", this._fetchSoundData.bind(this));
};

WaveForm.prototype._fetchSoundData = function(){
    var requestManager = new RequestManager();
    requestManager.uploadSound("test.wav", this.createWaveFormFromFile.bind(this));
};

WaveForm.prototype.createWaveFormFromFile = function(blob){
    this.waveform = WaveSurfer.create({
        container: "#waveform",
        waveColor: "#626262",
        progressColor: "#fff843"
    });
    this.waveform.loadBlob(blob);
};

WaveForm.prototype.createWaveForm = function(audioContext){
    this.waveform = WaveSurfer.create({
        audioContext: audioContext,
        container: "#waveform",
        waveColor: "#626262",
        progressColor: "#fff843"
    });
};
