// view/WaveForm.js
var inherit = require("../utils/inherit");
var AudioHelper = require("../utils/AudioHelper");
var BaseView = require("./BaseView");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("./windowsTransport");

module.exports = WaveForm;

function WaveForm(){
    BaseView.call(this, "wave-form");

    //$(this.getContainer()).attr("id", "waveform");
    //this.waveContainer = $("<div id='waveform'></div>");

    this.waveform = null;

    this._build();
}

inherit(WaveForm, BaseView);

WaveForm.prototype._build = function(){
};

WaveForm.prototype.create = function(buffer){
    this.waveform = WaveSurfer.create({
        container: this.getContainer().get(0),
        waveColor: "#626262",
        progressColor: "#fff843"
    });
    this.waveform.loadBlob(AudioHelper.AudioBufferToBlob(buffer));
    windowsTransport.subscribe(commonEventNames.E_PLAY_WAVE, this.waveform.playPause.bind(this.waveform));
    windowsTransport.subscribe(commonEventNames.E_STOP_WAVE, this.waveform.stop.bind(this.waveform));
};

WaveForm.prototype.createWaveForm = function(audioContext){
    this.waveform = WaveSurfer.create({
        container: this.getContainer().get(0),
        audioContext: audioContext,
        waveColor: "#626262",
        progressColor: "#fff843"
    });
    console.log(this.waveform);
};
