// model/AudioPlayer.js
var TrackManager = require("../utils/TrackManager");
var BaseTrackModel = require("./BaseTrackModel");
var ProjectModel = require("./ProjectModel");

module.exports = AudioPlayer;

function AudioPlayer(model){
    this.model = null;
    this.player = [];

    this.projectPlayer = null;

    this.setModel(model);
}

AudioPlayer.prototype.setModel = function(model){
    if(model){
        this.model = model;
        if(this.model instanceof ProjectModel){
            // merge before play
            this._fullPlayerWithProject();
        }
    }
};

AudioPlayer.prototype.play = function(){
    if(this.model instanceof BaseTrackModel){
        this.model.play();
    } else if(this.model instanceof ProjectModel){
        this.projectPlayer = new Tone.Player();
        this.projectPlayer.start();
    }
};

AudioPlayer.prototype._fullPlayerWithProject = function(){

};

AudioPlayer.prototype.stop = function(){
    if(this.model instanceof BaseTrackModel){
        this.model.stop();
    } else if(this.model instanceof ProjectModel){
        this.projectPlayer.stop();
    }
};
