// model/BaseTrackModel.js
var PostProcessSettings = require("./PostProcessSettings");
var generateUID = require("../core/generateUID");

module.exports = BaseTrackModel;

// @param {String} name
// @param {Object} source
function BaseTrackModel(id, data){
    this.id = id || -generateUID();
    this.length = data.length || 1;
    this.setting = data.setting || {};
    this.instrument = data.instrument || "synth";
    this.playSetting = data["play-setting"] || [];
    this.postSetting = data["post-setting"] || [];
    this.postProcessSettings = new PostProcessSettings(this.postSetting);
    this.trackObject = this._generate();
}

BaseTrackModel.prototype.createFormData = function(){
    var formData = new FormData();
    formData.append("user_audio", toBlob(this.trackObject), this.name);
    formData.append("postProcessSettings", JSON.stringify(this.postProcessSettings));
    return formData;
};

BaseTrackModel.prototype.getContext = function(){
    return this.trackObject.context._context;
};

BaseTrackModel.prototype.getConstants = function(){
    return this.trackObject.context;
};

BaseTrackModel.prototype.getAudioBuffer = function(){

};

/**
 * @param {Object} source
 * @type {null}
 * @protected
 * @return {Tone}
 */
BaseTrackModel.prototype._generate = null;

/**
 * @param {Object} options
 * @type {null}
 */
BaseTrackModel.prototype.play = null;

BaseTrackModel.prototype.playCompomponent = null;

BaseTrackModel.prototype.toJson = function(){
    return JSON.stringify(this);
};

BaseTrackModel.prototype.getData = function(){
    var result = {};
    result.id = this.id;
    result.instrument = this.instrument;
    result.length = this.length;
    result.setting = this.setting;
    result["play-setting"] = this.playSetting;
    result["post-setting"] = this.postSetting;
    return result;
};
