// model/DrumModel.js
module.exports = new DrumModel;

function DrumModel(){
    this.kickLeft = "kick-left";
    this.kickRight = "kick-right";
    this.bigTom = "big-tom";
    this.leftTom = "left-tom";
    this.bell = "bell";
    this.hiHat = "hi-hat";
    this.snare = "snare";
    this.hitHat1 = null;
    this.hitHat2 = null;
    this.hitHat3 = null;
    //this.keys = ['B', 'N', 'F', 'G', 'H', 'J', 'K', 'E', 'R', 'I'];
    this.keys = [['B', 'N'], ['F', 'G'], ['H', 'J'], ['T', 'Y']];
    this.images = ["/static/images/kick.png", "/static/images/tom.png",
                   "/static/images/snare.png", "/static/images/hi-hat.png"];
}

DrumModel.prototype.getDrumNameForKey = function(key){
    var result = null;
    switch(key){
        case 'B': result = this.kickLeft; break;
        case 'N': result = this.kickRight; break;
        case 'F': result = this.bigTom; break;
        case 'G': result = this.leftTom; break;
        case 'H': result = this.snare; break;
        case 'J': result = this.snare1; break;
        case 'T': result = this.bell; break;
        case 'Y': result = this.hiHat; break;
        default: console.log("wrong key: " + key); break;
    }
    return result;
};
