// model/FreeverbFilter.js
// TremoloFilter

module.exports = FreeverbFilter;

function FreeverbFilter(options){
    this.options = options || {};
}

FreeverbFilter.prototype.generate = function(){
    this.filterObject = new Tone.Freeverb(this.options);
};

FreeverbFilter.prototype.getDampening = function(){
    return this.filterObject.dampening.value;
};

FreeverbFilter.prototype.getRoomSize = function(){
    return this.filterObject.roomSize.value; // ???
};

FreeverbFilter.prototype.getWet = function(){
    return this.filterObject.wet.value; // ???
};

FreeverbFilter.prototype.getOptions = function(){
    var result = {};
    result.dampening = this.getDampening();
    result.roomSize = this.getRoomSize();
    result.wet = this.getWet();
    return result;
};

FreeverbFilter.prototype.setDampening = function(value){
    this.filterObject.dampening.value = value;
};

FreeverbFilter.prototype.setRoomSize = function(value){
    this.filterObject.roomSize.value = value; // ???
};

FreeverbFilter.prototype.setWet = function(value){
    this.filterObject.wet.value = value; // ???
};

FreeverbFilter.prototype.setOptions = function(options){
    this.setDampening(options.dampening);
    this.setRoomSize(options.roomSize);
    this.setWet(options.wet);
};
