// model/FreeverbFilter.js
var inherit = require("../utils/inherit");
var BaseFilterModel = require("./BaseFilterModel");

module.exports = FreeverbFilter;

function FreeverbFilter(options){
    BaseFilterModel.call(this);
    this.options = options || {};
}

inherit(FreeverbFilter, BaseFilterModel);

FreeverbFilter.prototype.generate = function(){
    this.filterObject = new Tone.Freeverb(this.options);
};

FreeverbFilter.prototype.getDampening = function(){
    return this.filterObject.dampening.value;
};

FreeverbFilter.prototype.getRoomSize = function(){
    return this.filterObject.roomSize.value; // ???
};

FreeverbFilter.prototype.getWet = function(){
    return this.filterObject.wet.value; // ???
};

FreeverbFilter.prototype.getOptions = function(){
    return this.options;
};

FreeverbFilter.prototype.setDampening = function(value){
    this.options.dampening = value;
    this.filterObject.dampening.value = value;
};

FreeverbFilter.prototype.setRoomSize = function(value){
    this.options.roomSize = value;
    this.filterObject.roomSize.value = value; // ???
};

FreeverbFilter.prototype.setWet = function(value){
    this.options.wet = value;
    this.filterObject.wet.value = value; // ???
};

FreeverbFilter.prototype.setOptions = function(options){
    this.setDampening(options.dampening);
    this.setRoomSize(options.roomSize);
    this.setWet(options.wet);
};

FreeverbFilter.prototype.setByName = function(optionName, value){
    switch(optionName){
        case "dampening":
            this.setDampening(value);
            break;
        case "roomSize":
            this.setRoomSize(value);
            break;
        case "wet":
            this.setWet(value);
            break;
    }
};
