// model/MixerRecorder.js
var inherit = require("../utils/inherit");
var BaseRecorder = require("./BaseRecorder");

module.exports = MixerRecorder;


function MixerRecorder(frequency, volume, startTime){
    BaseRecorder.call(this);
    this.frequency = frequency;
    this.volume = volume;
    this.startTime = startTime || 0;
    this.checkValues();
}

inherit(MixerRecorder, BaseRecorder);

MixerRecorder.prototype.checkValues = function(){
    var minFrequency = 60;
    var maxFrequency = 2000;
    var minVolume = -100;
    var maxVolume = 100;
    if(this.frequency < minFrequency){
        this.frequency = minFrequency;
    } else if(this.frequency > maxFrequency){
        this.frequency = maxFrequency;
    }
    if(this.volume < minVolume){
        this.volume = minVolume;
    } else if(this.volume > maxVolume) {
        this.volume = maxVolume;
    }
};

MixerRecorder.prototype.play = function(oscillator){
    var self = this;
    setTimeout(function(){
        oscillator.frequency.value = self.frequency;
        oscillator.volume.value = self.volume;
    }, this.startTime);
    oscillator.start('+' + (this.startTime / 1000));
};

MixerRecorder.prototype.getData = function(){
    var result = {};
    result.frequency = this.frequency;
    result.volume = this.volume;
    result.startTime = this.startTime;
    return result;
};
