// model/ProjectListModel.js
var ObservableList = require("./ObservableList");
var ProjectModel = require("./ProjectModel");
var inherit = require("../utils/inherit");
var commonEventNames = require("../core/commonEventNames");

module.exports = ProjectListModel;

function ProjectListModel(){
    ObservableList.call(this);
    this.__activeProject = null; // ProjectModel
}

inherit(ProjectListModel, ObservableList);

ProjectListModel.prototype.toString = function(){
    return "[object ProjectListModel]";
};

ProjectListModel.prototype.add = function(data){
    var projectModel = new ProjectModel(data);
    ObservableList.prototype.add.call(this, projectModel);
};

ProjectListModel.prototype.remove = function(projectId){
    var index = this.findIndexById(projectId);
    //var item = this.at(index);
    //item.isDeleted = true;
    if (index < this.size() && index >= 0) {
        // send signal to view for remove item from list
        this.observer.notify(commonEventNames.E_ITEM_REMOVED, index);
        // remove project from model
        this.__data.splice(index, 1);
    }
};

/**
 * @param {ProjectModel} project
 */
ProjectListModel.prototype.setActiveProject = function(project){
    this.activeProject = project;
};
