// model/ProjectListModel.js
var ObservableList = require("./ObservableList");
var TrackListModel = require("./ProjectModel");
var inherit = require("../utils/inherit");

module.exports = ProjectListModel;

function ProjectListModel(){
    ObservableList.call(this);
    this.__activeProject = null; // ProjectModel
}

inherit(ProjectListModel, ObservableList);

ProjectListModel.prototype.toString = function(){
    return "[object ProjectListModel]";
};

ProjectListModel.prototype.add = function(data){
    var trackListModel = new TrackListModel(data);
    ObservableList.prototype.add.call(this, trackListModel);
};

ProjectListModel.prototype.remove = function(projectId){
    var index = this.findIndexById(projectId);
    //var item = this.at(index);
    //item.isDeleted = true;
    ObservableList.prototype.remove.call(this, index);
    // remove project from model
    this.__data.splice(index, 1);
};

/**
 * @param {TrackListModel} project
 */
ProjectListModel.prototype.setActiveProject = function(project){
    this.activeProject = project;
};
