// model/ProxyTrackManager.js
// ProxyTrackManager

module.exports = {
    "updateSettingListFromTrack": function(settingList, track){ // SettingList OR FilterList
        updateList(settingList, track, setToSettingList)
    },
    "updateFilterListFromFilter": function(filterList, filter) {
        updateList(filterList, filter, setToFilterList);
        console.log(filterList);
    },
    "setSetting": function(list, track, settingName, value){
        setToTrackSetting(track, settingName, value);
        setToSettingList(list, track, settingName);
    },
    "setFilter": function(list, filter, filterName, value){

    }
};

/**
 *
 * @param listInstance - Array - SettingList OR FilterList
 * @param dataObject
 * @param callback
 */
function updateList(listInstance, dataObject, callback){
    var i;
    var name;
    var tokenSetting;
    var list = listInstance.list;
    for(i = 0; i < list.length; ++i) {
        tokenSetting = list[i]; // BaseTrackSetting type
        name = tokenSetting.name;
        callback(tokenSetting, dataObject, name);
    }
}

/**
 * Set all parameters from FilterObject to FilterList
 * @param filterElement
 * @param options
 */
function setOptionsToFilterList(filterElement, options){
    var optionName;
    for(optionName in options){
        filterElement.isEnabled = true; // switch flag for using this filter
        filterElement.options[optionName].set(options[optionName]);
    }
}

/**
 * Get data from track object and put them to SettingList
 * @param listElement
 * @param trackObject
 * @param optionName
 */
function setToSettingList(listElement, trackObject, optionName){
    switch(optionName){
        case "frequency":
            listElement.set(trackObject.getFrequency());
            break;
        case "volume":
            listElement.set( trackObject.getVolume());
            break;
        case "type":
            listElement.set(trackObject.getType());
            break;
        case "attack":
            listElement.set(trackObject.getAttack());
            break;
        case "decay":
            listElement.set(trackObject.getDecay());
            break;
        case "sustain":
            listElement.set(trackObject.getSustain());
            break;
        case "release":
            listElement.set(trackObject.getRelease());
            break;
    }
}

/**
 * Get data from filter object and put them to FilterList
 * @param listElement
 * @param filterObject
 * @param filterName
 */
function setToFilterList(listElement, filterObject, filterName){
    switch(filterName){
        case "tremolo":
            break;
        case "vibrato":
            break;
        case "crusher":
            break;
        case "phaser":
            break;
        case "freeverd":
            setOptionsToFilterList(listElement, filterObject);
            break;
    }
}

function setToTrackSetting(track, settingName, value){
    switch(settingName){
        case "frequency":
            track.setFrequency(value);
            break;
        case "volume":
            track.setVolume(value);
            break;
        case "type":
            track.setType(value);
            break;
        case "attack":
            track.setAttack(value);
            break;
        case "decay":
            track.setDecay(value);
            break;
        case "sustain":
            track.setSustain(value);
            break;
        case "release":
            track.setRelease(value);
            break;
    }
    track.setSetting(); // write to model parameter for saving to server
}

function setTockSetting(track, settingName, value){
    switch(settingName){
        case "frequency":
            track.setFrequency(value);
            break;
        case "volume":
            track.setVolume(value);
            break;
        case "type":
            track.setType(value);
            break;
        case "attack":
            track.setAttack(value);
            break;
        case "decay":
            track.setDecay(value);
            break;
        case "sustain":
            track.setSustain(value);
            break;
        case "release":
            track.setRelease(value);
            break;
    }
    track.setSetting(); // write to model parameter for saving to server
}
