// model/TrackDrum.js
// TrackDrum

var BaseTrackModel = require("./BaseTrackModel");
var inherit = require("../utils/inherit");
var KickLeft = require("./KickLeft");
var KickRight = require("./KickRight");
var TrackPlayer = require("./TrackPlayer");

module.exports = TrackDrum;

// @param {String} name
// @param {Object} source
function TrackDrum(id, data){
    BaseTrackModel.call(this, id, data);
    this.drums = data.drums || {};
    this.drumObjects = {};
    this.createDrumObjects();
}

inherit(TrackDrum, BaseTrackModel);

TrackDrum.prototype._generate = function(){
    return new Tone.Oscillator(this.setting).toMaster();
};

TrackDrum.prototype.createDrumObject = function(drum, playSettings){
    switch(drum){
        case "kick-left":
            this.drumObjects[drum] = new KickLeft(playSettings);
            break;
        case "kick-right":
            this.drumObjects[drum] = new KickRight(playSettings);
            break;
        case "":
            this.drumObjects[drum] = new TrackPlayer();
            break;
    }
};

TrackDrum.prototype.createDrumObjects = function(){
    var drum;
    for(drum in this.drums){
        this.createDrumObject(drum, this.drums[drum]["play-setting"]);
    }
};

TrackDrum.prototype.getData = function(){
    var result = {};
    result.instrument = this.instrument;
    result["post-setting"] = this.postSettings.getPostSettings();
    result["drums"] = this.getDrumObjectsData();
    return result;
};

TrackDrum.prototype.getDrumObjectsData = function(){
    var drum;
    var result = {};
    for(drum in this.drumObjects){
        result[drum] = this.drumObjects[drum].playSetting;
    }
    return result;
};

TrackDrum.prototype.emptyPlaySetting = function(){
    var drum;
    for(drum in this.drums){
        this.drums[drum].emptyPlaySetting();
    }
};

TrackDrum.prototype.createPlayObjects = function(){};
