// utils/RequestManager.js
var transportAudioFile = require("./transportAudioFile");
var TrackManager = require("./TrackManager");

var changeSound = require("../requests/changeSound");
var changeProject = require("../requests/changeProjectRequest");
var getSound = require("../requests/getSound");
var getProject = require("../requests/getProject");
var projectList = require("../requests/projectList");
var deleteProject = require("../requests/deleteProject");
var getUser = require("../requests/getUser");


module.exports = {
    "getUser": defineUser,
    "getProjectList": getProjectList,
    "deleteProject": deleteUserProject,
    "createSound": createSound,
    "createProject": addProject,
    "downloadSound": downloadSound,
    "uploadSound": uploadSound,
    "getProject": getProjectData
};

function defineUser(callback){
    getUser(callback);
    /*getUser(function(){
        //projectList("projects/", projectView.showProjectList.bind(projectView));
        projectList("projects/", callback.bind(context));
    });*/
}

function getProjectList(callback){
    projectList(callback);
}

function getProjectData(projectName, callback){
    var url = "projects/";
    var fullUrl = url + projectName + '/';
    getProject(fullUrl, callback);
}

function deleteUserProject(projectName, callback){
    var url = "projects/delete/";
    var fullUrl = url + projectName + '/';
    deleteProject(fullUrl, function(){
        console.log("delete " + projectName);
    });
}

function addProject(projectName, data, callback){
    var url = "projects/create/";
    var fullUrl = url + projectName + '/';
    changeProject(fullUrl, data, "json", callback);
}

function createSound(soundName, audioSrc){
    var url = "sounds/create/";
    // var url = "sound/update/";
    var fullUrl = url + soundName + '/';
    transportAudioFile(fullUrl, soundName, audioSrc, changeSound);
}

function downloadSound(soundName){
    var url = "sounds/download/";
    var fullUrl = url + soundName + '/';
    getSound(fullUrl, function(data){
        TrackManager.save(data, soundName);
    });
}

function uploadSound(soundName, callback){
    var url = "sounds/";
    var fullUrl = url + soundName + '/';
    getSound(fullUrl, callback);
}
