// utils/RequestManager.js
var transportAudioFile = require("./transportAudioFile");
var TrackManager = require("./TrackManager");

var changeSound = require("../requests/changeSound");
var getSound = require("../requests/getSound");
var getProject = require("../requests/getProject");
var projectList = require("../requests/projectList");
var deleteProject = require("../requests/deleteProject");
var getUser = require("../requests/getUser");

var MessageModal = require("../view/MessageModal");

module.exports = {
    "getUser": defineUser,
    "getProjectList": getProjectList,
    "deleteProject": deleteUserProject,
    "createSound": createSound,
    "downloadSound": downloadSound,
    "uploadSound": uploadSound,
    "getProject": getProjectData
};

function defineUser(callback){
    getUser(callback);
    /*getUser(function(){
        //projectList("projects/", projectView.showProjectList.bind(projectView));
        projectList("projects/", callback.bind(context));
    });*/
}

function getProjectList(callback){
    projectList(callback);
}

function getProjectData(projectName, callback){
    var url = "projects/";
    var fullUrl = url + projectName + '/';
    getProject(fullUrl, callback);
}

function deleteUserProject(projectName){
    var messageModal = new MessageModal();
    var url = "projects/delete/";
    var fullUrl = url + projectName + '/';
    deleteProject(fullUrl, messageModal.show.bind(this));
}

function createSound(soundName, audioSrc){
    var url = "sounds/create/";
    // var url = "sound/update/";
    var fullUrl = url + soundName + '/';
    transportAudioFile(fullUrl, soundName, audioSrc, changeSound);
}

function downloadSound(soundName){
    var url = "sounds/download/";
    var fullUrl = url + soundName + '/';
    getSound(fullUrl, function(data){
        TrackManager.save(data, soundName);
    });
}

function uploadSound(soundName, callback){
    var url = "sounds/";
    var fullUrl = url + soundName + '/';
    getSound(fullUrl, callback);
}
